### DISCUSSION TITLE
Private vs Protected definitions in core classes?

### DISCUSSION THREAD

**Discussion by prescience-data** [adonisjs/core#1945] [Help] [answered]
This is more a question for the team to avoid opening an issue, but I was wondering if the use of `private` as opposed to `protected` is a deliberate design choice for many of the core classes (such as `Application`).

Using `Application` as an example, the use of `private` on a method like `resolveModule` https://github.com/adonisjs/application/blob/dee840b0ce84bd7f220d3e4504c3656a10696517/src/Application.ts#L191 makes it impossible to extends the base class meaning it basically has to be hard forked to do any augmentation which seems unfortunate (as the developer misses any important new additions or upgrades). 

I completely understand why something like https://github.com/adonisjs/application/blob/dee840b0ce84bd7f220d3e4504c3656a10696517/src/Application.ts#L241 needs to `private` to avoid breaking specifically expected framework function, however something like https://github.com/adonisjs/application/blob/dee840b0ce84bd7f220d3e4504c3656a10696517/src/Application.ts#L275 seems like it should/could be `protected` to make it overridable. 

Would be great to get any insight from the maintainers around this! 😄 Thanks

**Comment by thetutlage** [comment, marked as answer]
May I know what you would want to achieve by over writing the `setEnvVars` method?

**Reply by prescience-data** [reply]
That was just the first example I found to use, but for instance:
```typescript

  /**
   * Sets env variables based upon the provided application info.
   */
  protected setEnvVars() {
    super.setEnvVars()
    // Extension:
    if (this.myExtendFrameworkNameVersion) {
      process.env.MY_EXENDED_FRAMEWORK_NAME_VERSION = this.myExtendFrameworkNameVersion.toString()
    }
  }


  /**
   * Loads the rc file from the application root
   */
  protected loadFooRcFile() {
    // Not really a good example, but a reason for needing resolveModule() exposed...
    return this.resolveModule("./.foo.json", () => {
      throw new Error('Foo Framework expects ".foo.json" file to exist in the application root')
    })
  }
```

**Reply by thetutlage** [reply]
Still I don't find much value in this, because then we will have to expose some API to load your Application class and not the default one.

**Reply by prescience-data** [reply]
In this context I am executing a custom Ignitor that calls the extended class.
_(as an aside, this `App` class would be nice to be able to extend and override rather than fork as well: https://github.com/adonisjs/core/blob/develop/src/Ignitor/Ace/App/index.ts)_
```typescript
import { Application } from "../core/Application"
import { ErrorHandler } from "@adonisjs/core/build/src/Ignitor/Ace/ErrorHandler"
import { SignalsListener } from "@adonisjs/core/build/src/Ignitor/SignalsListener"
//...
/**
 * Exposes the API to execute app commands registered under
 * the manifest file.
 */
export class App { 
// ...

 /**
   * Reference to the application
   */
  private application = new Application(this.appRoot, "console")

}

```

**Reply by thetutlage** [reply]
Yeah, this is not the use case I want to serve, coz there is no end to it. Same can be asked for Env, Config and 100 other internal classes. I know the change is small, but the motivation behind the change is not something I want to put labour for.

**Reply by prescience-data** [reply]
Understood, thanks for the feedback.
